File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/dynamic_programming/house_robber/P213_HouseRobberII.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        return max(self.dp2(nums, 0, len(nums) - 2), self.dp2(nums, 1, len(nums) - 1))

    def recursive(self, nums, start, end):
        if start >= end + 1:
            return 0
        # rob
        p1 = nums[start] + self.recursive(nums, start + 2, end)
        # not rob
        p2 = self.recursive(nums, start + 1, end)
        return max(p1, p2)

    def dp(self, nums, start, end):
        n = len(nums) + 2
        dp = [0 for i in range(n)]
        for i in range(end, start - 1, -1):
            dp[i] = max(nums[i] + dp[i + 2], dp[i + 1])
        return dp[start]

    def dp2(self, nums, start, end):
        p, p1, p2 = 0, 0, 0
        for i in range(end, start - 1, -1):
            p = max(nums[i] + p2, p1)
            p2 = p1
            p1 = p
        return p


if __name__ == '__main__':
    nums = [2, 3, 2]
    solu = Solution()
    print(solu.rob(nums))

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/dynamic_programming/house_robber/P213_HouseRobberII.java
package com.dsa.leetcode.labuladong.dynamic_programming.house_robber;

/**
 * @since 2023/11/30 22:16
 */
public class P213_HouseRobberII {

    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        return Math.max(dp2(nums, 0, nums.length - 2), dp2(nums, 1, nums.length - 1));
    }

    private int recursive(int[] nums, int start, int end) {
        if (start >= end + 1) {
            return 0;
        }
        // rob
        int p1 = nums[start] + recursive(nums, start + 2, end);
        // not rob
        int p2 = recursive(nums, start + 1, end);
        return Math.max(p1, p2);
    }

    private int dp(int[] nums, int start, int end) {
        int n = nums.length + 2;
        int[] dp = new int[n];
        // 从右到左
        for (int i = end; i >= start; i--) {
            dp[i] = Math.max(nums[i] + dp[i + 2], dp[i + 1]);
        }
        return dp[start];
    }

    private int dp2(int[] nums, int start,int end) {
        int p = 0;
        int p1 = 0;
        int p2 = 0;
        for (int i = end; i >= start; i--) {
            p = Math.max(nums[i] + p2, p1);
            p2 = p1;
            p1 = p;
        }
        return p;
    }

    public static void main(String[] args) {
        int[] nums = {1,2,3,1};
        P213_HouseRobberII p198 = new P213_HouseRobberII();
        System.out.println(p198.rob(nums));
    }
}
